### ISSUE TITLE
`npm:` specifiers still use version in lockfile when updating config

### ISSUE THREAD

**Issue by marvinhagemeister** [denoland/deno#20868]
I'm running into a bug with Fresh when trying to switch over to `npm:` specifiers. When using both `preact-render-to-string` and `preact` Deno loads multiple copies of Preact. This breaks preact's core `option` API which relies on a signleton.

It seems like anything I set in the import map is ignored. There I specifically request `preact@10.18.1`, but somehow `preact@10.15.1` is still getting installed by `preact-render-to-string`.

```json
{
  "vendor": true,
  "imports": {
    "preact": "npm:preact@10.18.1",
    "preact/": "npm:/preact@10.18.1/",
    "preact-render-to-string": "npm:preact-render-to-string@6.2.1"
  },
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "preact"
  }
}
```

<img width="328" alt="Screenshot 2023-10-10 at 13 39 15" src="https://github.com/denoland/deno/assets/1062408/ed121dd5-b22f-4139-a874-42a52d794be5">

## Steps to reproduce

1. Clone https://github.com/marvinhagemeister/deno-npm-version-bug
2. Run `deno run foo.tsx`

It will print `it works` when the bug is resolved, otherwise it will print `it doesn't work` if the bug is still present.

**Comment by nayeemrmn** [comment]
After deleting my `DENO_DIR`, `./deno.lock` and `./node_modules` it seems only `10.18.1` is included (though the repro program is still printing `it doesn't work`).

Before that I could reproduce, and saw the following in `deno.lock`:
```js
{
  "packages": {
    "specifiers": {
      "npm:preact-render-to-string@6.2.1": "npm:preact-render-to-string@6.2.1_preact@10.15.1",
      // ...
    }
    // ...
  }
}
```
Not sure what that mechanism is.

**Comment by marvinhagemeister** [comment]
That makes it sound like there is a stale cache somewhere. Maybe it stores the dependencies at the time the module was downloaded or something.

**Comment by dsherret** [comment]
preact-render-to-string is locked to 10.15.1:

https://github.com/marvinhagemeister/deno-npm-version-bug/blob/dbcaf16db6222d5f13dbd4b71b4870d0d265dfd2/deno.lock#L13

**Comment by marvinhagemeister** [comment]
Just ran into this again while trying to use more `npm:` specifiers in Fresh. It seems like this is a blocker for Fresh to be able to fully move over to them.

**Comment by dsherret** [comment]
I think when we see a specifier we haven't seen before, we should just invalidate the lockfile, but maintain the collection of npm specifiers to package name and version (not id with peer deps). I can look into this today.

**Comment by dsherret** [comment]
One other thing we could do is give precedence to npm specifiers in the import map. If one of them changes then we can assume that's the reference used in the project, remove it from the lockfile, then invalidate the other parts of the lockfile. That would require storing these specifiers in the lockfile though, but it would handle the scenario where you switch back and forth between a dep version.

**Comment by guillaume86** [comment]
Not sure it's the same issue but I'm trying to override a "ws" import inside the "puppeteer-core" package wich is a dependency of my direct dependency "npm:puppeteer@21.10.0".

Is this supposed to work? Or should I use npm: specifier somewhere? Or version specifier? 
```
  "imports": {
    "puppeteer": "npm:puppeteer@21.10.0",
    "ws": "./patched/ws.ts"
  },
```

**Comment by dsherret** [comment]
@guillaume86 no, that is https://github.com/denoland/deno/issues/18191

**Comment by luizfilipe** [comment]
I'm currently facing a similar issue on a try to override a dependency from a my project's dependency

I tried to use this, to fix an issue caused by the latest version of reestructure in my project, as per [docs ](https://docs.deno.com/runtime/manual/basics/import_maps#overriding-imports), but it does not seem to behave as expected

```
"scopes": {
  "fontkit@2.0.2": {
    "restructure": "npm:restructure@3.0.0"
  }
}
```

**Comment by dsherret** [comment]
Also, I think this issue original issue was essentially fixed by #22004. Going to close this now, but maybe let's open a targeted issue about anything remaining.
